lib/server.js

```diff
@@ -0,0 +1,237 @@
+/**
+ * @typedef {import('unist').Point} Point
+ * @typedef {import('unist').Position} UnistPosition
+ * @typedef {import('vfile-message').VFileMessage} VFileMessage
+ * @typedef {import('vscode-languageserver').Connection} Connection
+ * @typedef {import('vscode-languageserver-textdocument').TextDocument} TextDocument
+ * @typedef {import('vscode-languageserver').TextDocuments<TextDocument>} TextDocuments
+ * @typedef {import('..').Options} Options
+ */
+
+import {PassThrough} from 'node:stream'
+import {URL, pathToFileURL} from 'node:url'
+
+import {unified} from 'unified'
+import {engine} from 'unified-engine'
+import {VFile} from 'vfile'
+import {
+  Diagnostic,
+  DiagnosticSeverity,
+  Position,
+  Range,
+  TextDocumentSyncKind,
+  TextEdit
+} from 'vscode-languageserver/node.js'
+
+/**
+ * Convert a unist point to a language server protocol position.
+ *
+ * @param {Point} point
+ * @returns {Position}
+ */
+function unistPointToLspPosition(point) {
+  return Position.create(point.line - 1, point.column - 1)
+}
+
+/**
+ * @param {Point|null|undefined} point
+ * @returns {boolean}
+ */
+function isValidUnistPoint(point) {
+  return Boolean(
+    point && Number.isInteger(point.line) && Number.isInteger(point.column)
+  )
+}
+
+/**
+ * Convert a unist position to a language server protocol range.
+ *
+ * If no position is given, a range is returned  which represents the beginning
+ * of the document.
+ *
+ * @param {UnistPosition|null|undefined} position
+ * @returns {Range}
+ */
+function unistLocationToLspRange(position) {
+  if (position) {
+    if (isValidUnistPoint(position.start)) {
+      if (isValidUnistPoint(position.end)) {
+        return Range.create(
+          unistPointToLspPosition(position.start),
+          unistPointToLspPosition(position.end)
+        )
+      }
+
+      const start = unistPointToLspPosition(position.start)
+      return Range.create(start, start)
+    }
+
+    if (isValidUnistPoint(position.end)) {
+      const end = unistPointToLspPosition(position.end)
+      return Range.create(end, end)
+    }
+  }
+
+  return Range.create(0, 0, 0, 0)
+}
+
+/**
+ * Convert a vfile message to a language server protocol diagnostic.
+ *
+ * @param {VFileMessage} message
+ * @param {string|undefined} defaultSource
```

Comment:
is this parameter optional? (guessing from `undefined` being on the type)

```suggestion
 * @param {string|undefined} [defaultSource]
```

Reply:
`defaultSource` is an optional option for `configureUnifiedLanguageServer()`. At this internal function I don’t think it’s optional anymore. If you forget to pass it here, it’s probably a bug. It may be undefined though.

Reply:
I’m not entirely sure about `defaultSource` by the way. It could also be removed entirely.

Reply:
Better to remove it then I guess?